File: src/CircleAI.Tools/ComposioToolBridge.cs
// ComposioToolBridge.cs
//
// Routes B! tool calls to a Composio MCP server via JSON-RPC 2.0 over HTTP.
// Composio provides 250+ integrations (Gmail, Slack, GitHub, Calendar, etc.)
// through a single MCP endpoint.
// See: https://composio.dev/mcp

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Tools
{
    /// <summary>
    /// Routes tool calls to a Composio MCP server via JSON-RPC 2.0 over HTTP.
    /// Composio provides 250+ integrations (Gmail, Slack, GitHub, Calendar, etc.)
    /// through a single endpoint.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The bridge sends every tool invocation as a <c>tools/call</c> JSON-RPC 2.0
    /// request and interprets the response envelope to produce a <see cref="ToolResult"/>.
    /// </para>
    /// <para>
    /// Tool discovery calls <c>GET {serverUri}/tools</c> and maps each returned entry
    /// to a <see cref="ToolDefinition"/>.
    /// </para>
    /// </remarks>
    public sealed class ComposioToolBridge : IToolBridge, IDisposable, IAsyncDisposable
    {
        private static readonly Uri DefaultServerUri = new("https://mcp.composio.dev/", UriKind.Absolute);

        private readonly string _apiKey;
        private readonly Uri _serverUri;
        private readonly HttpClient _http;
        private readonly bool _ownsHttpClient;

        // JSON options shared across all serialisation calls.
        private static readonly JsonSerializerOptions JsonOptions = new()
        {
            PropertyNameCaseInsensitive = true,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        };

        /// <summary>
        /// Initialises the bridge.
        /// </summary>
        /// <param name="composioApiKey">Composio API key sent in the <c>X-API-Key</c> header.</param>
        /// <param name="serverUri">
        /// Base URI of the Composio MCP endpoint.
        /// Defaults to <c>https://mcp.composio.dev/</c>.
        /// </param>
        /// <param name="httpClient">
        /// Optional pre-configured <see cref="HttpClient"/>. When <c>null</c> a new instance
        /// is created and owned by this bridge (disposed in <see cref="DisposeAsync"/>).
        /// </param>
        public ComposioToolBridge(
            string composioApiKey,
            Uri? serverUri = null,
            HttpClient? httpClient = null)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(composioApiKey);

            _apiKey = composioApiKey;
            _serverUri = EnsureTrailingSlash(serverUri ?? DefaultServerUri);

            if (httpClient is not null)
            {
                _http = httpClient;
                _ownsHttpClient = false;
            }
            else
            {
                _http = new HttpClient();
                _ownsHttpClient = true;
            }

            _http.DefaultRequestHeaders.Accept.Clear();
            _http.DefaultRequestHeaders.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/json"));
        }

        // ------------------------------------------------------------------
        // IToolBridge — synchronous tool list (empty until discovery is run)
        // ------------------------------------------------------------------

        /// <summary>
        /// Synchronous available-tools list. Empty by default; call
        /// <see cref="GetAvailableToolsAsync"/> to populate via the Composio API.
        /// </summary>
        public IReadOnlyList<ToolDefinition> AvailableTools { get; private set; } =
            Array.Empty<ToolDefinition>();

        // ------------------------------------------------------------------
        // IToolBridge — invoke
        // ------------------------------------------------------------------

        /// <summary>
        /// Invokes a tool on the Composio MCP server via JSON-RPC 2.0.
        /// </summary>
        /// <exception cref="ArgumentException">
        /// Thrown when <paramref name="invocation"/> has a null or empty
        /// <see cref="ToolInvocation.ToolName"/>.
        /// </exception>
        public async Task<ToolResult> InvokeAsync(
            ToolInvocation invocation,
            CancellationToken ct = default)
        {
            ArgumentNullException.ThrowIfNull(invocation);
            if (string.IsNullOrWhiteSpace(invocation.ToolName))
                throw new ArgumentException("ToolName must not be null or whitespace.", nameof(invocation));

            var requestBody = new JsonRpcRequest
            {
                Method = "tools/call",
                Id     = 1,
                Params = new ToolCallParams
                {
                    Name      = invocation.ToolName,
                    Arguments = invocation.Arguments,
                },
            };

            var endpoint = new Uri(_serverUri, $"tools/{Uri.EscapeDataString(invocation.ToolName)}/invoke");

            try
            {
                using var request = BuildRequest(HttpMethod.Post, endpoint, requestBody);
                using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);

                var body = await response.Content
                    .ReadFromJsonAsync<JsonElement>(JsonOptions, ct)
                    .ConfigureAwait(false);

                if (!response.IsSuccessStatusCode)
                {
                    var httpError = $"HTTP {(int)response.StatusCode} {response.ReasonPhrase}";
                    return ToolResult.Failure(invocation.ToolName, ExtractError(body, httpError));
                }

                // Standard JSON-RPC 2.0 response: { "result": ..., "error": ... }
                if (body.TryGetProperty("error", out var errNode) &&
                    errNode.ValueKind != JsonValueKind.Null)
                {
                    var msg = errNode.TryGetProperty("message", out var m)
                        ? m.GetString() ?? errNode.ToString()
                        : errNode.ToString();
                    return ToolResult.Failure(invocation.ToolName, msg);
                }

                if (body.TryGetProperty("result", out var resultNode))
                    return ToolResult.Ok(invocation.ToolName, resultNode);

                // No result / error — treat as success with null payload.
                return ToolResult.Ok(invocation.ToolName);
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch (Exception ex)
            {
                return ToolResult.Failure(invocation.ToolName, ex.Message);
            }
        }

        // ------------------------------------------------------------------
        // IToolBridge — discovery
        // ------------------------------------------------------------------

        /// <summary>
        /// Fetches the list of tools available on the Composio MCP server
        /// (<c>GET {serverUri}/tools</c>) and caches it in <see cref="AvailableTools"/>.
        /// </summary>
        public async Task<IReadOnlyList<ToolDefinition>> GetAvailableToolsAsync(
            CancellationToken ct = default)
        {
            var endpoint = new Uri(_serverUri, "tools");

            try
            {
                using var request = BuildRequest(HttpMethod.Get, endpoint, body: null);
                using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);

                if (!response.IsSuccessStatusCode)
                    return Array.Empty<ToolDefinition>();

                var root = await response.Content
                    .ReadFromJsonAsync<JsonElement>(JsonOptions, ct)
                    .ConfigureAwait(false);

                var tools = ParseToolList(root);
                AvailableTools = tools;
                return tools;
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch
            {
                return Array.Empty<ToolDefinition>();
            }
        }

        // ------------------------------------------------------------------
        // IAsyncDisposable
        // ------------------------------------------------------------------

        /// <inheritdoc/>
        public void Dispose()
        {
            if (_ownsHttpClient)
                _http.Dispose();
        }

        /// <inheritdoc/>
        public ValueTask DisposeAsync()
        {
            Dispose();
            return ValueTask.CompletedTask;
        }

        // ------------------------------------------------------------------
        // Internal helpers
        // ------------------------------------------------------------------

        private HttpRequestMessage BuildRequest(HttpMethod method, Uri uri, object? body)
        {
            var req = new HttpRequestMessage(method, uri);
            req.Headers.TryAddWithoutValidation("X-API-Key", _apiKey);

            if (body is not null)
                req.Content = JsonContent.Create(body, options: JsonOptions);

            return req;
        }

        private static IReadOnlyList<ToolDefinition> ParseToolList(JsonElement root)
        {
            // Composio may return an array at root, or { "tools": [...] }.
            JsonElement toolsArray;
            if (root.ValueKind == JsonValueKind.Array)
            {
                toolsArray = root;
            }
            else if (root.TryGetProperty("tools", out var arr) && arr.ValueKind == JsonValueKind.Array)
            {
                toolsArray = arr;
            }
            else
            {
                return Array.Empty<ToolDefinition>();
            }

            var result = new List<ToolDefinition>(toolsArray.GetArrayLength());
            foreach (var item in toolsArray.EnumerateArray())
            {
                var name = item.TryGetProperty("name", out var n) ? n.GetString() : null;
                var desc = item.TryGetProperty("description", out var d) ? d.GetString() : string.Empty;

                if (string.IsNullOrWhiteSpace(name)) continue;

                var parameters = new Dictionary<string, ToolParameter>(StringComparer.Ordinal);
                var required   = new List<string>();

                if (item.TryGetProperty("inputSchema", out var schema) &&
                    schema.TryGetProperty("properties", out var props) &&
                    props.ValueKind == JsonValueKind.Object)
                {
                    foreach (var prop in props.EnumerateObject())
                    {
                        var type = prop.Value.TryGetProperty("type", out var t)
                            ? t.GetString() ?? "string"
                            : "string";
                        var propDesc = prop.Value.TryGetProperty("description", out var pd)
                            ? pd.GetString() ?? string.Empty
                            : string.Empty;

                        parameters[prop.Name] = new ToolParameter { Type = type, Description = propDesc };
                    }

                    if (schema.TryGetProperty("required", out var req) &&
                        req.ValueKind == JsonValueKind.Array)
                    {
                        foreach (var r in req.EnumerateArray())
                        {
                            var rName = r.GetString();
                            if (!string.IsNullOrWhiteSpace(rName)) required.Add(rName);
                        }
                    }
                }

                result.Add(new ToolDefinition
                {
                    Name               = name!,
                    Description        = desc ?? string.Empty,
                    Parameters         = parameters,
                    RequiredParameters = required,
                });
            }

            return result;
        }

        private static string ExtractError(JsonElement body, string fallback)
        {
            if (body.TryGetProperty("error", out var e))
            {
                if (e.TryGetProperty("message", out var m))
                    return m.GetString() ?? fallback;
                return e.ToString();
            }
            return fallback;
        }

        private static Uri EnsureTrailingSlash(Uri uri)
        {
            var s = uri.ToString();
            return s.EndsWith('/') ? uri : new Uri(s + "/", UriKind.Absolute);
        }

        // ------------------------------------------------------------------
        // Private DTO types (JSON-RPC 2.0)
        // ------------------------------------------------------------------

        private sealed class JsonRpcRequest
        {
            [JsonPropertyName("jsonrpc")]
            public string JsonRpc { get; } = "2.0";

            [JsonPropertyName("method")]
            public required string Method { get; init; }

            [JsonPropertyName("id")]
            public required int Id { get; init; }

            [JsonPropertyName("params")]
            public required object Params { get; init; }
        }

        private sealed class ToolCallParams
        {
            [JsonPropertyName("name")]
            public required string Name { get; init; }

            [JsonPropertyName("arguments")]
            public required IReadOnlyDictionary<string, object?> Arguments { get; init; }
        }
    }
}

File: python/src/circle_ai/tools/composio_tool_bridge.py
# composio_tool_bridge.py
#
# Port of CircleAI.Tools ComposioToolBridge.cs (C# — the EXACT spec).
#
# Routes B! tool calls to a Composio MCP server via JSON-RPC 2.0 over HTTP.
# Composio provides 250+ integrations (Gmail, Slack, GitHub, Calendar, etc.)
# through a single MCP endpoint. See: https://composio.dev/mcp
#
# The C# uses an injected HttpClient; the Python port injects the IHttpFetcher
# seam (circle_ai.integration.http) so the JSON-RPC + discovery logic is fully
# testable with InMemoryHttpFetcher and does no real network I/O.
#
# Porting notes:
#   * invoke_async sends a `tools/call` JSON-RPC 2.0 request to
#     `{server}/tools/{name}/invoke` and interprets the `{ "result", "error" }`
#     envelope into a ToolResult (result -> Ok; error.message -> Failure).
#   * get_available_tools_async does `GET {server}/tools` and maps each entry —
#     root array OR `{ "tools": [...] }` — to a ToolDefinition, reading
#     `inputSchema.properties` + `inputSchema.required`.
#   * The X-API-Key header is attached per request (mirrors
#     `TryAddWithoutValidation("X-API-Key", …)`); Accept is set to
#     application/json.
#   * Non-2xx responses return Failure with the JSON-RPC error message when
#     present, else "HTTP <code> <reason>".

from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..integration.http import HttpRequest, IHttpFetcher
from .tool_types import IToolBridge, ToolDefinition, ToolInvocation, ToolParameter, ToolResult

_DEFAULT_SERVER_URI = "https://mcp.composio.dev/"


class ComposioToolBridge(IToolBridge):
    """Routes tool calls to a Composio MCP server via JSON-RPC 2.0 over HTTP.
    Mirrors ``CircleAI.Tools.ComposioToolBridge``.
    """

    def __init__(
        self,
        composio_api_key: str,
        http_fetcher: IHttpFetcher,
        server_uri: Optional[str] = None,
    ) -> None:
        if composio_api_key is None or composio_api_key.strip() == "":
            raise ValueError("composio_api_key must be non-empty")
        if http_fetcher is None:
            raise ValueError("http_fetcher must not be None")

        self._api_key = composio_api_key
        self._http = http_fetcher
        base = server_uri if server_uri else _DEFAULT_SERVER_URI
        self._server_uri = base if base.endswith("/") else base + "/"
        self._available_tools: List[ToolDefinition] = []

    @property
    def available_tools(self) -> List[ToolDefinition]:
        """Synchronous available-tools list. Empty by default; call
        :meth:`get_available_tools_async` to populate via the Composio API.
        """
        return self._available_tools

    async def invoke_async(
        self,
        invocation: ToolInvocation,
        *,
        ct: Optional[object] = None,
    ) -> ToolResult:
        if invocation is None:
            raise ValueError("invocation must not be None")
        if invocation.tool_name is None or invocation.tool_name.strip() == "":
            raise ValueError("tool_name must not be null or whitespace")

        request_body = {
            "jsonrpc": "2.0",
            "method": "tools/call",
            "id": 1,
            "params": {
                "name": invocation.tool_name,
                "arguments": invocation.arguments,
            },
        }

        from urllib.parse import quote

        endpoint = self._server_uri + f"tools/{quote(invocation.tool_name, safe='')}/invoke"

        try:
            request = HttpRequest(
                method="POST",
                url=endpoint,
                headers=self._headers(),
                body_json=request_body,
            )
            response = await self._http.send_async(request)
            body = self._parse_json(response)

            if not response.is_success:
                http_error = f"HTTP {response.status_code} {response.reason}".rstrip()
                return ToolResult.failure(
                    invocation.tool_name, self._extract_error(body, http_error)
                )

            # Standard JSON-RPC 2.0 response: { "result": ..., "error": ... }
            if isinstance(body, dict):
                err = body.get("error")
                if err is not None:
                    if isinstance(err, dict) and err.get("message") is not None:
                        return ToolResult.failure(invocation.tool_name, str(err["message"]))
                    return ToolResult.failure(invocation.tool_name, str(err))

                if "result" in body:
                    return ToolResult.ok(invocation.tool_name, body["result"])

            # No result / error — treat as success with null payload.
            return ToolResult.ok(invocation.tool_name)
        except Exception as ex:  # noqa: BLE001 — mirror the C# catch-all.
            return ToolResult.failure(invocation.tool_name, str(ex))

    async def get_available_tools_async(
        self, *, ct: Optional[object] = None
    ) -> List[ToolDefinition]:
        """Fetch the list of tools available on the Composio MCP server
        (``GET {server}/tools``) and cache it in :attr:`available_tools`.
        Returns an empty list on any failure (mirrors the C# fallbacks).
        """
        endpoint = self._server_uri + "tools"
        try:
            request = HttpRequest(method="GET", url=endpoint, headers=self._headers())
            response = await self._http.send_async(request)
            if not response.is_success:
                return []
            root = self._parse_json(response)
            tools = self._parse_tool_list(root)
            self._available_tools = tools
            return tools
        except Exception:  # noqa: BLE001 — mirror the C# catch -> empty list.
            return []

    # ── Internal helpers ────────────────────────────────────────────────────

    def _headers(self) -> Dict[str, str]:
        return {"X-API-Key": self._api_key, "Accept": "application/json"}

    @staticmethod
    def _parse_json(response: object) -> Any:
        text = getattr(response, "text", "") or ""
        if text == "":
            return None
        import json as _json

        try:
            return _json.loads(text)
        except (ValueError, TypeError):
            return None

    @staticmethod
    def _parse_tool_list(root: Any) -> List[ToolDefinition]:
        # Composio may return an array at root, or { "tools": [...] }.
        if isinstance(root, list):
            tools_array = root
        elif isinstance(root, dict) and isinstance(root.get("tools"), list):
            tools_array = root["tools"]
        else:
            return []

        result: List[ToolDefinition] = []
        for item in tools_array:
            if not isinstance(item, dict):
                continue
            name = item.get("name")
            desc = item.get("description", "")

            if name is None or str(name).strip() == "":
                continue

            parameters: Dict[str, ToolParameter] = {}
            required: List[str] = []

            schema = item.get("inputSchema")
            if isinstance(schema, dict):
                props = schema.get("properties")
                if isinstance(props, dict):
                    for prop_name, prop_val in props.items():
                        if isinstance(prop_val, dict):
                            ptype = prop_val.get("type") or "string"
                            pdesc = prop_val.get("description") or ""
                        else:
                            ptype, pdesc = "string", ""
                        parameters[prop_name] = ToolParameter(type=str(ptype), description=str(pdesc))

                    req = schema.get("required")
                    if isinstance(req, list):
                        for r in req:
                            if r is not None and str(r).strip() != "":
                                required.append(str(r))

            result.append(
                ToolDefinition(
                    name=str(name),
                    description=str(desc) if desc is not None else "",
                    parameters=parameters,
                    required_parameters=required,
                )
            )

        return result

    @staticmethod
    def _extract_error(body: Any, fallback: str) -> str:
        if isinstance(body, dict):
            err = body.get("error")
            if isinstance(err, dict):
                if err.get("message") is not None:
                    return str(err["message"])
                return str(err)
            if err is not None:
                return str(err)
        return fallback
